src/EditorFeatures/CSharp/EventHookup/EventHookupSessionManager_EventHookupSession.cs

```diff
@@ -106,27 +107,21 @@ public EventHookupSession(
 
             var asyncToken = asyncListener.BeginAsyncOperation(GetType().Name + ".Start");
 
-            _eventNameTask = Task.Factory.SafeStartNewFromAsync(
-                () => DetermineIfEventHookupAndGetHandlerNameAsync(document, position, cancellationToken),
-                cancellationToken,
-                TaskScheduler.Default);
+            _eventNameTask = DetermineIfEventHookupAndGetHandlerNameAsync(document, position, cancellationToken);
 
-            var continuedTask = _eventNameTask.SafeContinueWithFromAsync(
-                async t =>
+            ContinueOnMainThreadAsync(_eventNameTask).CompletesAsyncOperation(asyncToken);
+
+            return;
+
+            async Task ContinueOnMainThreadAsync(Task<string?> eventNameTask)
+            {
+                var eventName = await eventNameTask.ConfigureAwait(true);
+                if (eventName != null)
                 {
                     await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(alwaysYield: true, cancellationToken);
-
-                    // Once we compute the name, update the tooltip (if we haven't already been dismissed)
-                    if (this._eventNameTask != null && t.Result != null)
-                    {
-                        commandHandler.EventHookupSessionManager.EventHookupFoundInSession(this, t.Result);
-                    }
-                },
-                cancellationToken,
-                TaskContinuationOptions.OnlyOnRanToCompletion | TaskContinuationOptions.ExecuteSynchronously,
-                TaskScheduler.Default);
-
-            continuedTask.CompletesAsyncOperation(asyncToken);
+                    commandHandler.EventHookupSessionManager.EventHookupFoundInSession(this, eventName);
```

Comment:
>commandHandler.EventHookupSessionManager.EventHookupFoundInSession(this, eventName); [](http://example.com/codeflow?start=20&length=84)

I'm not clear why this needs to be so complicated. Instead of doing the Continue* code, couldn't DetermineIfEventHookupAndGetHandlerNameAsync do a SwithToMainThread when it's ready to return and then do the EventHookupFoundInSession directly?